### PR TITLE
Parkobj loads rct_gx

### PR DESCRIPTION
**Pull request by spacek531**
Loading of sprite atlases is not as fast as desired. rct_gx archives seem to load much faster.

Test objects: https://github.com/OpenRCT2/OpenRCT2/discussions/16746#discussioncomment-2298523

Will address https://github.com/OpenRCT2/OpenRCT2/issues/15547

### CODE REVIEW

File: src/openrct2/drawing/Drawing.h
Comment (duncanspumpkin) [inline src/openrct2/drawing/Drawing.h:520]: this is not required to be exposed.

File: src/openrct2/drawing/Drawing.Sprite.cpp
Comment (duncanspumpkin) [inline src/openrct2/drawing/Drawing.Sprite.cpp:365]: pass the buffer by const reference.
Name the function GfxLoadGx

File: src/openrct2/drawing/Drawing.Sprite.cpp
Comment (duncanspumpkin) [inline src/openrct2/drawing/Drawing.Sprite.cpp:382]: I would perform this within the ImageTable. Once my PR is merged that will handle this bit.

File: src/openrct2/object/ImageTable.cpp
Comment (duncanspumpkin) [inline src/openrct2/object/ImageTable.cpp:144]: you are missing the `if (rangeStart != std::string::npos)`

File: src/openrct2/drawing/Drawing.Sprite.cpp
Comment (spacek531) [reply src/openrct2/drawing/Drawing.Sprite.cpp:382]: Do you want me to move the code to ImageTable or remove it from the PR entirely?

File: src/openrct2/object/ImageTable.cpp
Comment (duncanspumpkin) [inline src/openrct2/object/ImageTable.cpp:242]: After my PR this part of the function will change. This will use common code from `LoadObjectImages`

File: src/openrct2/drawing/Drawing.Sprite.cpp
Comment (duncanspumpkin) [reply src/openrct2/drawing/Drawing.Sprite.cpp:382]: move it for now.

File: src/openrct2/drawing/Drawing.Sprite.cpp
Comment (spacek531) [inline src/openrct2/drawing/Drawing.Sprite.cpp:386]:
```suggestion
        return std::optional(gx);
```
a does not compile on linux, b does not compile on windows

Comment (duncanspumpkin) [reply src/openrct2/drawing/Drawing.Sprite.cpp:386]: you can just do `{ gx }` but tbh it shouldcompile fine wirth `return gx`

Comment (spacek531) [reply src/openrct2/drawing/Drawing.Sprite.cpp:386]: windows does not like `{ gx }` either: 'no instance of constructor "std::optional<_Ty>::optional [with _Ty=rct_gx]" matches the arguments list
argument types are: (rct_gx)

Comment (duncanspumpkin) [reply src/openrct2/drawing/Drawing.Sprite.cpp:386]: `std::make_optional(std::move(gx));`

**Comment by spacek531** [comment]
will close https://github.com/OpenRCT2/OpenRCT2/issues/15547

**Comment by spacek531** [comment]
test objects available here: https://github.com/OpenRCT2/OpenRCT2/discussions/16746#discussioncomment-2298523

**Comment by Gymnasiast** [comment]
> will close #15547

Not directly - it will provide a way to close it, but we would still need to actually convert the images to gx and ship a new objects release.

File: src/openrct2/drawing/Drawing.h
Comment (duncanspumpkin) [inline src/openrct2/drawing/Drawing.h:529]: I read this so many times and somehow never noticed gfx is spelt wrong. GfxLoadGx

**Review by duncanspumpkin — APPROVED** [review]

**Comment by spacek531** [comment]
Should a change like this get a changelog entry? I am thinking yes

**Comment by duncanspumpkin** [comment]
One area of this that is still under discussion is do we actually need a range should it just be a path.

**Comment by IntelOrca** [comment]
@duncanspumpkin allow both

**Comment by spacek531** [comment]
$DAT:ObjData/ does not allow a path without a range does it?

**Comment by duncanspumpkin** [comment]
> $DAT:ObjData/ does not allow a path without a range does it? I'm fine with either, I am wondering if there is precedent.

There's a slight issue in that objects specify an identifier where as you should be specifying a path i wonder perhaps if the json should be reorged to make this more clear.

**Comment by spacek531** [comment]
I suppose it is easier to know how many images are in an rct_gx than an object, considering there is ride object import code specifically to guess how many images there are

**Comment by duncanspumpkin** [comment]
Just looking at the json we are supporting we will now have the following:
```javascript
    "images": [
       "$LGX:spritefile-rle.lgx[0..4970]"
     ],
```
and 
```javascript
    "images": [
       "$LGX:spritefile-rle.lgx"
    ],
```
which will be added to the following supported formats:
```javascript
    "images": [
          { "path": "images/preview.png","x": 0, "y": 0 },
          { "path": "images/0.png","x": 0, "y": 0 },
          etc...
    ],

    "images": [
        { "path": "preview.png", "x": 1, "y": 1 },
        "$CSG[44289..44339]"
    ],

    "images": [
        { "path": "preview.png", "x": 1, "y": 1 },
        "$RCT2:OBJDATA/PATHDIRT.DAT[73..145]"
    ],

    "images": [
        "$G1[22856..22871]",
        "$G1[22565..22576]"],
```

**Review by Gymnasiast — APPROVED** [review]
Seems to work as intended. I have **not** reviewed the code, I just tested it.

@duncanspumpkin I leave the final decision to you (in case you want some changes still).

**Comment by spacek531** [comment]
> Just looking at the json we are supporting we will now have the following:

Is it good or bad to have all these formats? They all seem necessary to me except lgx-no-range

**Comment by duncanspumpkin** [comment]
> > Just looking at the json we are supporting we will now have the following:
> 
> Is it good or bad to have all these sources? Since they all use the same range format, I don't see the issue - the odd one out is lgx without range. I'm fine with dropping lgx without range because it's the only format that does not explicitly define how many sprites there are, but I can see it as a tiny qol feature to not have to write the range when using a generation tool.

I wrote that message because initially i felt like lgx would be out of place. After writing out all the formats I started to warm to it. I don't think there is anything left to discuss on this and am happy to merge. 

I personally think we should only use bmp (nonrle) with this but that will be a discussion to be had within the objects repo.

**Pull request merged**
